lib/server.js

```diff
@@ -789,51 +774,78 @@ Server.prototype._handle = function _handle(req, res) {
  */
 Server.prototype._route = function _route(req, res, name, cb) {
     var self = this;
+    // helper function to, when on router error, emit error events and then
+    // flush the err.
+    var errResponse = function errResponse(err) {
+        return self._emitErrorEvents(req, res, null, err, function () {
+            res.send(err);
```

Comment:
all `res.send()` requires to check `res.headersSent` otherwise node will throw `Can't set headers after they are sent.`